src/sentry/ratelimits/cardinality.py

```diff
@@ -132,3 +157,266 @@ def use_quotas(
+            set_counts = dict(zip(set_keys_to_count, results))
+
+        grants = []
+        cardinality_sample_factor = self._get_set_cardinality_sample_factor()
+        for request in requests:
+            if not request.quotas:
+                grants.append(
+                    GrantedQuota(
+                        request=request, granted_unit_hashes=request.unit_hashes, reached_quotas=[]
+                    )
+                )
+                continue
+
+            granted_hashes = []
+
+            quotas_by_remaining_limit = defaultdict(list)
+            for quota in request.quotas:
+                remaining_limit = max(
+                    0,
+                    quota.limit
+                    - cardinality_sample_factor
+                    * sum(
+                        set_counts[k] for k in self._get_read_sets_keys(request, quota, timestamp)
+                    ),
+                )
+                quotas_by_remaining_limit[remaining_limit].append(quota)
+
+            # Determine the quota(s) with the smallest remaining limit
+            # (quota.limit - <set count from redis>). If we drop any hashes at
+            # all, we can only have hit the most restrictive quota, so we
+            # report those as `reached_quotas`.
+            smallest_remaining_limit = min(quotas_by_remaining_limit)
+            smallest_remaining_limit_running = smallest_remaining_limit
+            reached_quotas = []
+
+            # for each hash in the request, check if:
+            # 1. the hash is in `unit_keys`. If so, it has already been seen in
+            #    this timewindow, and ingesting additional copies of it comes
+            #    at no cost (= ingesting multiple metric buckets of the same
+            #    timeseries only counts once against quota)
+            #
+            # 2. we still have budget/"remaining_limit". In that case,
+            #    accept/admit the hash as well and reduce the remaining quota.
+            #
+            # 3. we have observed a totally new hash. in that case set
+            #    `reached_quotas` for reporting purposes, but don't add the
+            #    hash to `granted_hashes` (which is our return value)
+            for hash in request.unit_hashes:
+                if unit_keys[self._get_timeseries_key(request, hash)]:
+                    granted_hashes.append(hash)
+                elif smallest_remaining_limit_running > 0:
+                    granted_hashes.append(hash)
+                    smallest_remaining_limit_running -= 1
+                else:
+                    reached_quotas = quotas_by_remaining_limit[smallest_remaining_limit]
+
+            grants.append(
+                GrantedQuota(
+                    request=request,
+                    granted_unit_hashes=granted_hashes,
+                    reached_quotas=reached_quotas,
+                )
+            )
+
+        return timestamp, grants
+
+    def use_quotas(
```

Comment:
I think we have a problem with this abstraction. 
If we want to move `use_quotas` in a separate thread we will want to write the current time window on the main thread and only write the other time windows (plus maybe the TTL) in the separate thread. Otherwise updating will be delayed enough that race conditions will become extremely frequent.

So, since we are not yet running the multi threaded approach, this is good. But we let's call it out in a TODO so we do not forget about it.

Reply:
ok, where do we want to bump the TTL of the `cardinality:timeseries` keys? main thread?